lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -537,53 +561,119 @@ def from_relation_data(cls, data: dict):
             model=data["model"],
             model_uuid=data["model_uuid"],
             application=data["application"],
-            charm_name=data["charm_name"],
+            unit=data.get("unit", ""),
+            charm_name=data.get("charm_name", ""),
         )
 
     @property
     def identifier(self) -> str:
         """Format the topology information into a terse string."""
-        return "{}_{}_{}".format(self.model, self.model_uuid, self.application)
-
-    @property
-    def scrape_identifier(self):
-        """Format the topology information into a scrape identifier."""
-        return "juju_{}_{}_{}_prometheus_scrape".format(
-            self.model,
-            self.model_uuid[:7],
-            self.application,
-        )
+        # This is odd, but may have `None` as a model key
+        return "_".join([str(val) for val in self.as_dict().values()])
 
     @property
     def promql_labels(self) -> str:
         """Format the topology information into a verbose string."""
-        return 'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
-            self.model, self.model_uuid, self.application
+        return ", ".join(
+            [
+                'juju_{}="{}"'.format(key, value)
+                for key, value in self.as_dict(rename_keys={"charm_name": "charm"}).items()
+            ]
         )
 
-    def as_dict(self) -> dict:
+    def as_dict(self, rename_keys: Optional[Dict[str, str]] = None) -> dict:
         """Format the topology information into a dict."""
-        return {
+        ret = {
             "model": self.model,
             "model_uuid": self.model_uuid,
             "application": self.application,
+            "unit": self.unit,
             "charm_name": self.charm_name,
         }
 
-    def as_dict_with_promql_labels(self):
+        ret["unit"] or ret.pop("unit")
```

Comment:
My initial process in going through this was that there were/are already too many derived classes present, which make the mental model for any of this library extremely challenging, especially with names like `ProviderTopology` and not `JujuTopologyFromProxy`.

If anything, I would destroy `AggregatorTopology` and `ProviderTopology`, and simply provide additional `@classmethod` functions on `JujuTopology` itself, with a `__getattr__` method to look at a `self.topology` dict, which shortcuts all of this.

Then:

```
class JujuTopology:
    def __init__(self, topology: dict = None):
        ...
    def __getattr__(self, key):
        try:
            return self.topology[key]
        except KeyError:
            raise AttributeError(key)
    def __setattr__(self, key, value):
        self.topology[key] = value
```

No more need for conversions with `as_dict` (just return `self.topology`), no more lines like the one above, etc.

That said, that kind of metaprogramming is overkill, and I think we're overthinking things here. The responsibility of `JujuTopology` is to provide topology information. As little as is required to uniquely identify something, and as much as possible thanks to @mmanciop's approval of "unit and charm at the same time is ok".

We can't remove things, but we **can** add things and not break anything at all other than marginally higher usage from TSDB (and even with the added topology label, it's still less than the number of labels the [official Prometheus operator](https://github.com/prometheus-operator/prometheus-operator) adds to things).

The mental model for consumers is much clearer without a kaleidoscope of single-purpose classes.

To put it differently, what is the expected harm from adding an additional label if it's available? The aggregator [did not previously include charm names](https://github.com/canonical/prometheus-k8s-operator/pull/187/files#diff-3eb4f4fb71ab351956c8e7b96b07a804995e81540a0d2e94c2f63333c7792f79L1880) (and still doesn't), but this is worse IME.

Users/administrators of the aggregator who `juju relate ...` don't have a good reason to expect that, when going through the aggregator, they should expect `juju_unit` but without, they should get `juju_charm`. It's the same observability stack, and they should get the same labels (or as close as we can make them) without arbitrary distinctions which require reading a bunch of docs so they can write appropriate alert rules/etc.

This is, to me, a sane default -- to try our best to make the labels from different sources overlap rather than diverge.

Reply:
Okay. If multiple topology classes (an implementation of [Template Method Pattern](https://en.wikipedia.org/wiki/Template_method_pattern)) is deemed a complex mental model then let us consider our options of having a single Topology class.

1. Firstly it may help if we have a shared understanding on what we are trying to achieve by this refactor (particularly in view, my note below regarding nested groups bug fixed in scrape config charm). I would be grateful for your input and what is our objective with this refactor ?

2. Even if we want to have only a single Juju Topology class that does not necessarily mean we design this classes constructor such that it declares multiple optional arguments and then raises an exception if one of them is missing (cf. discussion above regarding this). Here are some alternative we could consider in developing such a single Juju Topology class

- The constructor may only accept the arguments that are always guaranteed to be available.
- Additional class methods (factory functions) can be provided (actually already are) that accept the additional arguments (unit name and charm name) and generate the class that behaves in conformity to the arguments provided.